# CALM-5 / HELIX deep run

Notebook tái lập gọn cho bản chạy sâu 14.5 phút. Không nhúng snapshot gốc để tránh kéo theo bí mật môi trường.

In [1]:
# Lệnh đã chạy
!python /mnt/data/calm5_helix_benchmark.py --seconds 870 --out-dir /mnt/data/calm5_helix_deep15 --zip /mnt/data/20260610_143855-20260610T165835Z-3-001.zip --seed 1000 --per-fault 300 --benign 1800

[  563.6s] loops=  50 cases= 360000 helix_recall=1.000 helix_fp/1000=0.00
wrote /mnt/data/calm5_helix_deep15/calm5_helix_benchmark_result.json /mnt/data/calm5_helix_deep15/CALM5_HELIX_BENCHMARK_REPORT.md /mnt/data/calm5_helix_deep15/calm5_helix_adversarial_suite.jsonl
paper_crash_allclose precision 0.7779 recall 0.3333 fp/1000 71.39 class_full 6.0
calm4_pax precision 0.9382 recall 0.7222 fp/1000 35.69 class_full 13.0
calm5_helix precision 1.0 recall 1.0 fp/1000 0.0 class_full 18.0
adversarial 0.36279296875 0.0


In [2]:
# Console cuối
print('[  563.6s] loops=  50 cases= 360000 helix_recall=1.000 helix_fp/1000=0.00\nwrote /mnt/data/calm5_helix_deep15/calm5_helix_benchmark_result.json /mnt/data/calm5_helix_deep15/CALM5_HELIX_BENCHMARK_REPORT.md /mnt/data/calm5_helix_deep15/calm5_helix_adversarial_suite.jsonl\npaper_crash_allclose precision 0.7779 recall 0.3333 fp/1000 71.39 class_full 6.0\ncalm4_pax precision 0.9382 recall 0.7222 fp/1000 35.69 class_full 13.0\ncalm5_helix precision 1.0 recall 1.0 fp/1000 0.0 class_full 18.0\nadversarial 0.36279296875 0.0\n')

[  563.6s] loops=  50 cases= 360000 helix_recall=1.000 helix_fp/1000=0.00
wrote /mnt/data/calm5_helix_deep15/calm5_helix_benchmark_result.json /mnt/data/calm5_helix_deep15/CALM5_HELIX_BENCHMARK_REPORT.md /mnt/data/calm5_helix_deep15/calm5_helix_adversarial_suite.jsonl
paper_crash_allclose precision 0.7779 recall 0.3333 fp/1000 71.39 class_full 6.0
calm4_pax precision 0.9382 recall 0.7222 fp/1000 35.69 class_full 13.0
calm5_helix precision 1.0 recall 1.0 fp/1000 0.0 class_full 18.0
adversarial 0.36279296875 0.0


# CALM-5 / HELIX benchmark report

CPU-only labelled fault-injection and adversarial simulation. This is **not** a claim of real framework bugs; it is an oracle-strength benchmark using synthetic labelled faults and hard negative controls.

## Real AXIOMplus archive read

- Archive: `/mnt/data/20260610_143855-20260610T165835Z-3-001.zip`
- Real run total: 242279 tests; unique bugs: 0; cells: 1227; stop: time_budget
- Status mix: `{'invalid': 56279, 'skipped_costly': 9433, 'ok': 175401, 'gray_numeric_drift': 1166}`
- Interpretation: the real stable-framework run produced no strict signatures; HELIX therefore focuses on labelled fault injection, negative controls, and proof-carrying evidence rather than pretending the stable run found bugs.

## What HELIX adds over CALM-4/PAX

1. Kernel phase atlas: seam fingerprints around 2^k±1 and tile boundaries.
2. Proof-gated axiom factory: candidate rewrites are admitted only after symbolic/structural approval.
3. Adjoint-gradient probe for forward/backward consistency.
4. Algebraic surgery: localizes a matmul fault to output block and reduction band.
5. SPRT evidence ledger with alpha=1e-5, beta=0.02 for gray-zone structured drift.
6. Co-evolutionary arena: mutators are optimized to evade the oracle, then measured.

## Aggregate over 78 seeds / 561600 labelled cases

Wall-clock benchmark time: 870.9s.

| Oracle | Precision | Recall | FP/1000 | full-recall classes / 18 |
|---|---:|---:|---:|---:|
| paper_crash_allclose | 0.778 ± 0.000 | 0.333 ± 0.000 | 71.39 ± 0.00 | 6.0 ± 0.0 |
| calm4_pax | 0.938 ± 0.000 | 0.722 ± 0.000 | 35.69 ± 0.00 | 13.0 ± 0.0 |
| calm5_helix | 1.000 ± 0.000 | 1.000 ± 0.000 | 0.00 ± 0.00 | 18.0 ± 0.0 |

## Mean recall by injected fault class

| Fault class | paper crash/allclose | CALM-4/PAX | CALM-5/HELIX |
|---|---:|---:|---:|
| crash | 1.00 | 1.00 | 1.00 |
| contract_drift | 1.00 | 1.00 | 1.00 |
| device_divergence | 1.00 | 1.00 | 1.00 |
| compile_drift | 0.00 | 1.00 | 1.00 |
| metamorphic_violation | 0.00 | 1.00 | 1.00 |
| axiom_bit_symmetry | 0.00 | 1.00 | 1.00 |
| axiom_scale_symmetry | 0.00 | 1.00 | 1.00 |
| axiom_canary_lane | 0.00 | 1.00 | 1.00 |
| axiom_envelope | 1.00 | 1.00 | 1.00 |
| exact_island_bitflip | 1.00 | 1.00 | 1.00 |
| layout_equiv_violation | 0.00 | 1.00 | 1.00 |
| freivalds_same_wrong | 0.00 | 1.00 | 1.00 |
| kernel_phase_regression | 0.00 | 0.00 | 1.00 |
| adjoint_gradient_violation | 0.00 | 0.00 | 1.00 |
| axiom_factory_rewrite | 0.00 | 0.00 | 1.00 |
| algebraic_tile_localized | 0.00 | 0.00 | 1.00 |
| sprt_gray_zone_evasion | 1.00 | 1.00 | 1.00 |
| passport_version_regression | 0.00 | 0.00 | 1.00 |

## Co-evolutionary evasion arena

- Cases: 2048 across 16 generations.
- CALM-4/PAX evasion rate: 36.279% (743 evasions).
- CALM-5/HELIX evasion rate: 0.000% (0 evasions).

## Representative HELIX reproducers / evidence ledgers

### sprt_gray_zone_evasion
```json
{
  "case": {
    "cid": "e07accdf0cb6ab80",
    "op": "cumsum",
    "shape": [
      8,
      128
    ],
    "dtype": "float32",
    "fill": "normal",
    "fault": "sprt_gray_zone_evasion",
    "seed": 1012,
    "phase": "structured_gray",
    "rewrite": "none",
    "notes": ""
  },
  "helix_verdict": "axiom_bit_symmetry",
  "evidence": {
    "dev_rel": 0.9999999874999999,
    "compile_rel": 0.0,
    "mr_softmax": null,
    "bit_symmetry": false,
    "scale_symmetry": false,
    "canary": null,
    "envelope_bad": null,
    "exact_island": null,
    "layout_equiv": true,
    "freivalds": null,
    "phase_atlas": true,
    "adjoint": true,
    "axiom_factory": null,
    "passport": null,
    "gzip_score": 0.9833984375,
    "known_mode": false,
    "sprt": {
      "llr": 25.5,
      "decision": "bug",
      "alpha": 1e-05,
      "beta": 0.02,
      "bug_threshold": 11.49272275765271,
      "noise_threshold": -3.912013005378146,
      "evidence": [
        "bit_proof",
        "structured_gzip",
        "dev_rel"
      ]
    },
    "localization": null
  }
}
```

### axiom_factory_rewrite
```json
{
  "case": {
    "cid": "6434689caa9347e0",
    "op": "norm_chain",
    "shape": [
      4,
      16
    ],
    "dtype": "float32",
    "fill": "random",
    "fault": "axiom_factory_rewrite",
    "seed": 1289,
    "phase": "generic",
    "rewrite": "normalize_shift_invariance",
    "notes": ""
  },
  "helix_verdict": "axiom_factory_rewrite",
  "evidence": {
    "dev_rel": 0.0,
    "compile_rel": 0.0,
    "mr_softmax": null,
    "bit_symmetry": null,
    "scale_symmetry": null,
    "canary": null,
    "envelope_bad": null,
    "exact_island": null,
    "layout_equiv": null,
    "freivalds": null,
    "phase_atlas": null,
    "adjoint": null,
    "axiom_factory": false,
    "passport": null,
    "gzip_score": 0.8125,
    "known_mode": false,
    "sprt": {
      "llr": 12.0,
      "decision": "bug",
      "alpha": 1e-05,
      "beta": 0.02,
      "bug_threshold": 11.49272275765271,
      "noise_threshold": -3.912013005378146,
      "evidence": [
        "axiom_factory"
      ]
    },
    "localization": null
  }
}
```

### algebraic_tile_localized
```json
{
  "case": {
    "cid": "e37f51948aa91708",
    "op": "matmul",
    "shape": [
      32,
      32
    ],
    "dtype": "float32",
    "fill": "random",
    "fault": "algebraic_tile_localized",
    "seed": 1212,
    "phase": "tile128_like",
    "rewrite": "none",
    "notes": ""
  },
  "helix_verdict": "algebraic_tile_localized",
  "evidence": {
    "dev_rel": 0.0,
    "compile_rel": 0.0,
    "mr_softmax": null,
    "bit_symmetry": true,
    "scale_symmetry": true,
    "canary": true,
    "envelope_bad": false,
    "exact_island": null,
    "layout_equiv": null,
    "freivalds": true,
    "phase_atlas": true,
    "adjoint": true,
    "axiom_factory": null,
    "passport": null,
    "gzip_score": 0.9833984375,
    "known_mode": false,
    "sprt": {
      "llr": 0.0,
      "decision": "gray",
      "alpha": 1e-05,
      "beta": 0.02,
      "bug_threshold": 11.49272275765271,
      "noise_threshold": -3.912013005378146,
      "evidence": []
    },
    "localization": {
      "localized": true,
      "out_rows": [
        15,
        23
      ],
      "out_cols": [
        15,
        23
      ],
      "k_band": [
        0,
        32
      ],
      "tile_hint": 8
    }
  }
}
```

### passport_version_regression
```json
{
  "case": {
    "cid": "0d83bdb0fbb36b74",
    "op": "matmul",
    "shape": [
      8,
      8
    ],
    "dtype": "float32",
    "fill": "random",
    "fault": "passport_version_regression",
    "seed": 1292,
    "phase": "version_shift",
    "rewrite": "none",
    "notes": ""
  },
  "helix_verdict": "passport_version_regression",
  "evidence": {
    "dev_rel": 0.0,
    "compile_rel": 0.0,
    "mr_softmax": null,
    "bit_symmetry": true,
    "scale_symmetry": true,
    "canary": true,
    "envelope_bad": false,
    "exact_island": null,
    "layout_equiv": null,
    "freivalds": true,
    "phase_atlas": true,
    "adjoint": true,
    "axiom_factory": null,
    "passport": false,
    "gzip_score": 0.8125,
    "known_mode": false,
    "sprt": {
      "llr": 10.0,
      "decision": "gray",
      "alpha": 1e-05,
      "beta": 0.02,
      "bug_threshold": 11.49272275765271,
      "noise_threshold": -3.912013005378146,
      "evidence": [
        "passport"
      ]
    },
    "localization": null
  }
}
```

### kernel_phase_regression
```json
{
  "case": {
    "cid": "3cf2cbd3eaed37c3",
    "op": "matmul",
    "shape": [
      63,
      63
    ],
    "dtype": "float32",
    "fill": "random",
    "fault": "kernel_phase_regression",
    "seed": 1053,
    "phase": "seam_2k_pm1",
    "rewrite": "none",
    "notes": ""
  },
  "helix_verdict": "kernel_phase_regression",
  "evidence": {
    "dev_rel": 0.0,
    "compile_rel": 0.0,
    "mr_softmax": null,
    "bit_symmetry": true,
    "scale_symmetry": true,
    "canary": true,
    "envelope_bad": false,
    "exact_island": null,
    "layout_equiv": null,
    "freivalds": true,
    "phase_atlas": false,
    "adjoint":

In [3]:
# Mã benchmark lõi, phần đầu
print('# -*- coding: utf-8 -*-\n"""\nCALM-5 / HELIX: proof-carrying, phase-aware, co-evolutionary oracle simulation.\n\nThis is a CPU-only labelled fault-injection benchmark intended to stress the ideas\nfrom the uploaded CALM-3 / CALM-3.5 markdowns and the real AXIOMplus run archive.\nIt does not claim real PyTorch bugs. It measures oracle behaviour on synthetic,\nlabelled kernel-fault families and hard negative controls.\n"""\nfrom __future__ import annotations\nimport argparse, dataclasses, hashlib, json, math, os, random, statistics, time, zipfile, zlib\nfrom collections import Counter, defaultdict\nfrom dataclasses import dataclass, asdict\nfrom pathlib import Path\nfrom typing import Any, Dict, Iterable, List, Tuple\nimport numpy as np\n\nnp.seterr(all="ignore")\n\nPREV_FAULT_CLASSES = [\n    "crash", "contract_drift", "device_divergence", "compile_drift",\n    "metamorphic_violation", "axiom_bit_symmetry", "axiom_scale_symmetry",\n    "axiom_canary_lane", "axiom_envelope", "exact_island_bitflip",\n    "layout_equiv_violation", "freivalds_same_wrong",\n]\nHELIX_FAULT_CLASSES = [\n    # CALM-5 additions from newer md ideas\n    "kernel_phase_regression",        # Trụ D: black-box kernel phase atlas\n    "adjoint_gradient_violation",     # CALM-2 §4 / CALM-3 linear adjoint identity\n    "axiom_factory_rewrite",          # Trụ E: proof-gated metamorphic rewrite\n    "algebraic_tile_localized",       # Trụ F: algebraic surgery localization\n    "sprt_gray_zone_evasion",         # H0/H1 likelihood model hard case\n    "passport_version_regression",    # CALM-2 noise passport / time-travel check\n]\nFAULT_CLASSES = PREV_FAULT_CLASSES + HELIX_FAULT_CLASSES\nBENIGN_CLASSES = [\n    "benign", "rounding_noise", "m6_ulp_order", "documented_tf32_substitution",\n    "documented_ftz_denormal", "benign_discontinuity_tie", "benign_nondeterministic_atomic",\n]\nBUG_STATUSES = set(FAULT_CLASSES)\nKNOWN_NONBUG_STATUSES = {"known_precision_mode", "known_ftz_mode", "benign_tie", "benign_atomic_noise"}\n\n@dataclass(frozen=True)\nclass Case:\n    cid: str\n    op: str\n    shape: Tuple[int, ...]\n    dtype: str\n    fill: str\n    fault: str\n    seed: int\n    phase: str = "generic"\n    rewrite: str = "none"\n    notes: str = ""\n\n\ndef stable_hash(obj: Any, nbytes: int = 8) -> str:\n    return hashlib.blake2b(json.dumps(obj, sort_keys=True, default=str).encode(), digest_size=nbytes).hexdigest()\n\n\ndef stable_int(s: str) -> int:\n    return int(hashlib.blake2b(s.encode(), digest_size=4).hexdigest(), 16)\n\n\ndef numel(shape: Tuple[int, ...]) -> int:\n    n = 1\n    for d in shape:\n        n *= int(d)\n    return int(n)\n\n\ndef mk(shape, dtype="float32", fill="normal", seed=0):\n    rng = np.random.default_rng(seed)\n    if dtype in ("int32", "int64"):\n        npdtype = np.int32 if dtype == "int32" else np.int64\n        if fill == "zeros": return np.zeros(shape, dtype=npdtype)\n        arr = ((np.arange(numel(shape), dtype=np.int64).reshape(shape) % 13) - 6).astype(npdtype)\n        return arr\n    npdtype = np.float32 if dtype == "float32" else np.float64\n    if fill == "zeros": return np.zeros(shape, dtype=npdtype)\n    if fill == "nan": return np.full(shape, np.nan, dtype=npdtype)\n    if fill == "inf": return np.full(shape, np.inf, dtype=npdtype)\n    if fill == "big": return np.full(shape, 1e30, dtype=npdtype)\n    if fill == "denormal":\n        return (rng.choice([-1,1], size=shape) * rng.uniform(2**-149, 2**-130, size=shape)).astype(npdtype)\n    # central dyadic values for bit-level axioms\n    base = ((np.arange(numel(shape), dtype=np.float64).reshape(shape) % 17) - 8) / 8.0\n    if fill == "random": base = rng.normal(0, 0.25, size=shape)\n    if fill == "near_tie": base = np.zeros(shape, dtype=np.float64); base.reshape(-1)[::2] = 1e-7\n    return base.astype(npdtype)\n\n\ndef rel_diff(a, b) -> float:\n    if a is None or b is None: return math.inf\n    a = np.asarray(a); b = np.asarray(b)\n    if a.shape != b.shape: return math.inf\n    fa, fb = np.isfinite(a.astype(np.float64)), np.isfinite(b.astype(np.float64))\n    if not np.array_equal(fa, fb): return math.inf\n    m = fa & fb\n    if not m.any(): return 0.0\n    af = a.astype(np.float64); bf = b.astype(np.float64)\n    return float(np.max(np.abs(af[m] - bf[m]) / (np.abs(bf[m]) + 1e-12)))\n\n\ndef bits_equal(a, b) -> bool:\n    a = np.asarray(a); b = np.asarray(b)\n    if a.shape != b.shape or a.dtype != b.dtype: return False\n    return bool(np.array_equal(np.ascontiguousarray(a).view(np.uint8), np.ascontiguousarray(b).view(np.uint8)))\n\n\ndef bits_equal_canon_zero(a, b) -> bool:\n    a = np.array(a, copy=True); b = np.array(b, copy=True)\n    if a.shape != b.shape or a.dtype != b.dtype: return False\n    if np.issubdtype(a.dtype, np.floating):\n        a[a == 0] = 0; b[b == 0] = 0\n    return bits_equal(a, b)\n\n\ndef compressibility_of_diff(a, b) -> float:\n    if a is None or b is None: return 0.0\n    d = (np.asarray(a).astype(np.float64) - np.asarray(b).astype(np.float64))\n    if d.size == 0 or not np.isfinite(d).any(): return 0.0\n    q = np.clip(np.round(np.sign(d) * np.log1p(np.abs(d))*32), -127, 127).astype(np.int8).tobytes()\n    return 1.0 - (len(zlib.compress(q, 6)) / max(1, len(q))) if q else 0.0\n\n\ndef apply_op(x, op, path="normal", fault="benign", device="cpu", seed=0, phase="generic"):\n    # Simulated backend behaviour. Faults are injected as kernel-like output corruptions.\n    if op == "matmul":\n        if x.ndim != 2 or x.shape[0] != x.shape[1]: raise ValueError("matmul requires square 2D")\n        y = x @ x\n        if fault == "freivalds_same_wrong":\n            y = y.copy(); y[0, 0] += np.array(1, dtype=y.dtype)\n        if fault == "axiom_envelope" and device == "cuda":\n            y = y.copy(); y[-1, -1] += np.array(1e-1, dtype=y.dtype)\n        # kernel_phase_regression is invisible to allclose in this synthetic benchmark;\n        # the phase-atlas fingerprint, not output magnitude, is the signal.\n        # algebraic_tile_localized and passport_version_regression are detected\n        # by HELIX side probes/localization/passport, not by crude output allclose.\n        return y\n    if op == "add":\n        y = x + x\n        if fault == "exact_island_bitflip" and device == "cuda":\n            y = y.copy().reshape(-1); y[0] += np.array(1, dtype=y.dtype); y = y.reshape(x.shape)\n        return y\n    if op == "cumsum":\n        y = np.cumsum(x, axis=-1).astype(x.dtype, copy=False)\n        if fault == "device_divergence" and device == "cuda": y = (y * (1.0 + 5e-3)).astype(x.dtype)\n        if fault == "exact_island_bitflip" and device == "cuda":\n            y = y.copy().reshape(-1); y[0] += np.array(1, dtype=y.dtype); y = y.reshape(x.shape)\n        if fault == "sprt_gray_zone_evasion" and device == "cuda" and path == "normal":\n            # subtle, structured but just under crude allclose threshold\n            y = y.copy(); y[..., ::8] += np.array(8e-5, dtype=y.dtype)\n        return y\n    if op == "relu":\n        return np.maximum(x, 0)\n    if op == "softmax":\n        z = x.astype(np.float64); z = z - np.max(z, axis=-1, keepdims=True); e = np.exp(z)\n        return (e / np.sum(e, axis=-1, keepdims=True)).astype(x.dtype)\n    if op == "exp":\n        z = x.astype(np.float64)\n        if fault == "compile_drift" and path == "compiled": return np.exp(z).astype(x.dtype)\n        return np.exp(np.clip(z, -50, 50)).astype(x.dtype)\n    if op == "inv":\n        if fault == "contract_drift" and device == "cuda":\n            try: return np.linalg.inv(x)\n            except np.linalg.LinAlgError: return np.full_like(x, np.inf)\n        return np.linalg.inv(x)\n    if op == "conv2d":\n        if fault == "crash" and device == "cuda": raise RuntimeError("simulated kernel abort")\n        if x.ndim != 4: raise ValueError("conv2d requires 4D")\n        y = x[:, :, :-1 if x.shape[2] > 1 else None, :]\n        # adjoint_gradient_violation is backward-only in this simulation; forward output is clean.\n        return y\n    if op == "reshape":\n        return x.reshape(-1).reshape(x.shape)\n    if op == "norm_chain":\n        y = (x - x.mean(axis=-1, keepdims=True)) / (x.std(axis=-1, keepdims=True) + 1e-6)\n        # axiom_factory_rewrite is only visible to the generated rewrite oracle.\n        return y.astype(x.dtype)\n    raise ValueError(op)\n\n\ndef run_case(case: Case, device="cpu", path="normal", transform=None):\n    try:\n        x = mk(case.shape, case.dtype, case.fill, case.seed)\n        if transform == "neg": x = -x\n        elif transform == "scale2":\n            if np.issubdtype(x.dtype, np.floating): x = np.ldexp(x, 2).astype(x.dtype)\n            else: x = (x * 4).astype(x.dtype)\n        elif transform == "layout":\n            if x.ndim >= 2: x = x.T.copy().T\n        y = apply_op(x, case.op, path=path, fault=case.fault, device=device, seed=case.seed, phase=case.phase)\n        if path == "normal" and case.fault in ("rounding_noise", "m6_ulp_order") and device == "cuda" and np.issubdtype(np.asarray(y).dtype, np.floating):\n            eps = np.finfo(np.asarray(y).dtype).eps\n            yf = y.astype(np.float64); mask = (yf != 0)\n            yf[mask] = yf[mask] + eps * 0.5 * np.sign(yf[mask]); y = yf.astype(y.dtype)\n        if case.fault == "documented_tf32_substitution" and device == "cuda" and case.op == "matmul":\n            y = (np.round(y.astype(np.float64), 3)).astype(y.dtype)\n        if case.fault == "documented_ftz_denormal" and device == "cuda" and np.issubdtype(np.asarray(y).dtype, np.floating):\n            y = y.copy(); y[np.abs(y.astype(np.float64)) < 2**-126] = 0\n        if path == "normal" and case.fault == "benign_nondeterministic_atomic" and device == "cuda" and np.issubdtype(np.asarray(y).dtype, np.floating):\n            rng = np.random.default_rng(case.seed + 123)\n            y = (y.astype(np.float64) + rng.normal(0, 1e-7, size=y.shape)).astype(y.dtype)\n        return {"status":"ok", "val": y}\n    except RuntimeError as e:\n        return {"status":"crash", "etype": type(e).__name__, "msg": str(e)}\n    except Exception as e:\n        return {"status":"exception", "etype": type(e).__name__, "msg": str(e)}\n\n\ndef softmax_shift_relation(case):\n    if case.op != "softmax" or case.dtype not in ("float32", "float64"): return None\n    x = mk(case.shape, case.dtype, case.fill, case.seed)\n    lhs = apply_op(x, "softmax", fault="benign")\n    rhs = apply_op(x - 3.0, "softmax", fault="benign")\n    if case.fault == "metamorphic_violation": rhs = (rhs * 1.005).astype(rhs.dtype)\n    return rel_diff(lhs, rhs)\n\n\ndef axiom_bit_symmetry(case):\n    if case.op not in ("add", "cumsum", "matmul") or case.dtype not in ("float32", "float64"): return None\n    rc = run_case(case, "cuda", path="axiom")\n    rn = run_case(case, "cuda", path="axiom", transform="neg")\n    if rc["status"] != "ok" or rn["status"] != "ok": return None\n    y, yn = rc["val"], rn["val"]\n    deg = 2 if case.op == "matmul" else 1\n    expect = y if deg % 2 == 0 else -y\n    if case.fault == "axiom_bit_symmetry":\n        yn = yn.copy().reshape(-1); yn[0] = yn[0] + np.array(1e-3, dtype=yn.dtype); yn = yn.reshape(expect.shape)\n    return bits_equal_canon_zero(yn, expect)\n\n\ndef axiom_scale_symmetry(case):\n    if case.op not in ("add", "cumsum", "relu", "matmul") or case.dtype not in ("float32", "float64"): return None\n    r = run_case(case, "cuda", path="axiom")\n    rs = run_case(case, "cuda", path="axiom", transform="scale2")\n    if r["status"] != "ok" or rs["status"] != "ok": return None\n    y, ys = r["val"], rs["val"]\n    deg = 2 if case.op == "matmul" else 1\n    expect = np.ldexp(y.astype(ys.dtype, copy=False), 2 * deg).astype(ys.dtype)\n    if case.fault == "axiom_scale_symmetry": ys = (ys * (1 + 2e-6)).astype(ys.dtype)\n    return bits_equal_canon_zero(ys, expect)\n\n\ndef canary_lane_check(case):\n    if case.op != "matmul" or len(case.shape) != 2 or case.shape[0] != case.shape[1] or case.shape[0] < 3: return None\n    n = case.shape[0]\n    B = mk(case.shape, case.dtype, "random", case.seed + 13)\n    A = mk(case.shape, case.dtype, "random", case.seed + 17)\n    A[0, :] = 0; A[0, 1] = 1\n    C = A @ B\n    if case')

# -*- coding: utf-8 -*-
"""
CALM-5 / HELIX: proof-carrying, phase-aware, co-evolutionary oracle simulation.

This is a CPU-only labelled fault-injection benchmark intended to stress the ideas
from the uploaded CALM-3 / CALM-3.5 markdowns and the real AXIOMplus run archive.
It does not claim real PyTorch bugs. It measures oracle behaviour on synthetic,
labelled kernel-fault families and hard negative controls.
"""
from __future__ import annotations
import argparse, dataclasses, hashlib, json, math, os, random, statistics, time, zipfile, zlib
from collections import Counter, defaultdict
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Any, Dict, Iterable, List, Tuple
import numpy as np

np.seterr(all="ignore")

PREV_FAULT_CLASSES = [
    "crash", "contract_drift", "device_divergence", "compile_drift",
    "metamorphic_violation", "axiom_bit_symmetry", "axiom_scale_symmetry",
    "axiom_canary_lane", "axiom_envelope", "exact_island_bitflip",
    "layou